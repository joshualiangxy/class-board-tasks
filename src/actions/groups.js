import firebase from '../firebase/firebase';
import { v4 as uuid } from 'uuid';
import { addGroup } from './user';

export const newGroup = (name, groupId, module) => ({
  type: 'NEW_GROUP',
  name,
  groupId,
  module
});

// Should return a promise of both adding a user to the group and adding a group to a user.
export const startNewGroup = (groupName, module) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const gid = uuid();
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(gid);

    const userPromise = userRef.update({
      groups: firebase.firestore.FieldValue.arrayUnion(gid)
    });
    const groupPromise = groupRef
      .set({
        name: groupName,
        module
      })
      .catch(error => {
        console.log(error);
      });
    const groupUserPromise = groupRef.collection('users').doc(uid).set({
      uid: uid,
      displayName: getState().user.displayName,
      studentNum: getState().user.studentNum,
      admin: true
    });

    return Promise.all([userPromise, groupPromise, groupUserPromise]).then(
      () => {
        dispatch(newGroup(groupName, gid, module));
        dispatch(addGroup(gid));
      }
    );
  };
};

export const addNewUser = (user, gid) => {
  // Note: No checking on this side, since only valid users can be added by getUser
  // Add user to group
  const groupPromise = firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(user.uid)
    .set({
      displayName: user.displayName,
      studentNum: user.studentNum,
      uid: user.uid,
      admin: false
    });
  // Add group to user's groups array.
  const userPromise = firebase
    .firestore()
    .collection('users')
    .doc(user.uid)
    .update({
      groups: firebase.firestore.FieldValue.arrayUnion(gid)
    });
  return Promise.all([groupPromise, userPromise]);
};

export const setGroups = groups => ({ type: 'SET_GROUPS', groups });

export const startSetGroups = () => {
  const groups = []; // The array of objects to be pushed
  return (dispatch, getState) => {
    firebase
      .firestore()
      .collection('users')
      .doc(getState().auth.user.uid)
      .get()
      .then(userSnapshot => {
        // Get the array of all groups this user is in, and dispatch it to the store.
        const groupIds = userSnapshot.data().groups || []; // Array of group id, or empty array.
        const promises = [];

        groupIds.forEach(gid => {
          promises.push(
            firebase
              .firestore()
              .collection('groups')
              .doc(gid)
              .get()
              .then(snapshot => {
                groups.push({
                  name: snapshot.get('name'),
                  gid,
                  module: snapshot.get('module') || ''
                });
              })
          );
        });

        return Promise.all(promises);
      })
      .then(() => dispatch(setGroups(groups)))
      .catch(error => console.log(error));
  };
};

export const leaveGroup = gid => ({
  type: 'LEAVE_GROUP',
  gid
});

export const startLeaveGroup = (gid, count) => {
  console.log(count);
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const groupRef = firebase.firestore().collection('groups').doc(gid);

    // Promises
    const userPromise = firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .update({
        groups: firebase.firestore.FieldValue.arrayRemove(gid)
      }); // Remove this group from the user's groups.
    const groupPromise = groupRef.collection('users').doc(uid).delete(); // Remove the user document from this group
    return Promise.all([userPromise, groupPromise])
      .then(() => {
        if (count <= 1) {
          // This is for the case where the user is alone, and only this case, so there shouldn't be
          // any risk of deleting a group someone else is in. Unless someone decides to multi-tab this to break the system.
          // However, this problem can also be solved if we do an additional database-check to ensure there aren't any users left.
          // The count includes the current user.

          // Deletes all of the documents under the users collection.
          const usersPromise = groupRef.collection('users').get().then(snapshot => {
            snapshot.forEach(doc => doc.ref.delete())
          })
            // Deletes all of the documents under the tasks collection.
          const tasksPromise = groupRef.collection('tasks').get().then(snapshot => {
            snapshot.forEach(doc => doc.ref.delete())
          })
          return Promise.all([usersPromise, tasksPromise]).then(() => groupRef.delete());
        } else {
          return;
        }
      })
      .then(() => dispatch(leaveGroup(gid)));
  };
};

export const getUser = email => {
  // Get UID from email
  return firebase
    .firestore()
    .collection('emailToUid')
    .doc(email)
    .get()
    .then(uidDoc => {
      if (uidDoc.exists) {
        // If this user exists, find the UID associated
        return firebase
          .firestore()
          .collection('users')
          .doc(uidDoc.data().uid)
          .get()
          .then(userDoc => {
            // Then return an object containing the user's name and number
            return {
              displayName: userDoc.data().displayName,
              studentNum: userDoc.data().studentNum,
              uid: userDoc.ref.id
            };
          });
      } else {
        return undefined;
      }
    });
};

// Not to be confused with getUser, this one is for all the users in a group.
export const getAllUsers = gid => {
  return () =>
    firebase
      .firestore()
      .collection('groups')
      .doc(gid)
      .collection('users')
      .get()
      .then(query => {
        const result = [];
        query.forEach(doc => {
          result.push(doc.data());
        });
        return result;
      });
};

export const kickUser = (user, gid) => {
  // user is a user object, group is gid
  // Since users aren't stored inside the store, no need to dispatch
  const uid = user.uid;
  // Remove this user from the group's collection of users
  const groupPromise = firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(uid)
    .delete();
  // Remove this group from the user's groups
  const userPromise = firebase
    .firestore()
    .collection('users')
    .doc(uid)
    .update({
      groups: firebase.firestore.FieldValue.arrayRemove(gid)
    });
  // Remove the user from any tasks 
  const tasksPromise = firebase.firestore().collection('tasks').get().then(query => {
    // Not fully implemented yet. 
    // TODO: This should remove the person's field from the completed object of the doc.
    query.forEach(doc => {
      const update = {};

      update[`completed.${uid}`] = firebase.firestore.FieldValue.delete();
      doc.update(update);
    })
  })
  return Promise.all([groupPromise, userPromise]);
};

export const promoteUser = (user, gid) => {
  const uid = user.uid;
  return firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(uid)
    .update({
      admin: true
    });
};
export const demoteUser = (user, gid) => {
  const uid = user.uid;
  return firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(uid)
    .update({
      admin: false
    });
};