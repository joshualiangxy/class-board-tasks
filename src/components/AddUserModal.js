import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
