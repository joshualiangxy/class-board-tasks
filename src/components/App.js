import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

