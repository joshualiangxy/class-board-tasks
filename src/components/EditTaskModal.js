import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import {
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks
} from '../actions/tasks';
import TaskForm from './TaskForm';

export const EditTaskModal = ({
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  task,
  users,
  editGroupTask
}) => {
  const submitTask = updates =>
    gid
      ? startEditGroupTask(task.id, updates, groupName).then(() => {
          editGroupTask(task.id, updates);
          startSetTasks(); // Temporary solution
          onRequestClose();
        })
      : startEditPersonalTask(task.id, updates).then(() => onRequestClose());

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Edit Task"
      onRequestClose={onRequestClose}
      appElement={document.getElementById('root')}
    >
      <h2>Edit Task</h2>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
        task={task}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startEditPersonalTask: (id, updates) =>
    dispatch(startEditPersonalTask(id, updates)),
  startEditGroupTask: (id, updates, groupName) =>
    dispatch(startEditGroupTask(id, updates, groupName)),
  startSetTasks: () => dispatch(startSetTasks())
});

export default connect(undefined, mapDispatchToProps)(EditTaskModal);
