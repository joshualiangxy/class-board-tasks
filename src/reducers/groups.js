const groupsReducer = (state = [], action) => {
  switch (action.type) {
    case 'NEW_GROUP':
      const newGroup = { name: action.name, gid: action.groupId, module: action.module };
      return [...state, newGroup];
    case 'SET_GROUPS':
      return action.groups;
    case 'LEAVE_GROUP':
      return state.filter(group => group.gid !== action.gid);
    default:
      return state;
  }
};

export default groupsReducer;
