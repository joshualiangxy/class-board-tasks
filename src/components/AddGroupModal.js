import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
