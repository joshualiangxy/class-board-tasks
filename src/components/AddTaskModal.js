import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
