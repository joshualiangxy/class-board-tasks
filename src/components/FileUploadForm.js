import React, { useState } from 'react';
import firebase from '../firebase/firebase';
import { startUpdateDownloadURL } from '../actions/tasks';
import { uploadFile, startRemoveUserFile } from '../actions/files';
import { connect } from 'react-redux';

export const FileUploadForm = ({
  id,
  uid,
  uploadFile,
  TaskEvent,
  startUpdateDownloadURL,
  startRemoveUserFile,
  fileExists
}) => {
  const [error, setError] = useState('');
  const [uploading, setUploading] = useState(false);

  const fileUpload = e => {
    e.preventDefault();
    e.persist();

    const element = document.getElementById(id);
    const files = element.files;
    if (files.length > 0) {
      const file = files[0];

      Promise.resolve()
        .then(() => {
          if (fileExists) return startRemoveUserFile(uid);
        })
        .then(() => {
          const uploadTask = uploadFile(file);

          uploadTask.on(
            TaskEvent.STATE_CHANGED,
            () => setUploading(true),
            error => {
              switch (error.code) {
                case 'storage/unauthorized':
                  setError('Not allowed to upload file!');
                  break;
                case 'storage/canceled':
                  setError('File upload cancelled');
                  break;
                default:
                  setError('File upload failed');
                  break;
              }
              setUploading(false);
            },
            () => {
              const ref = uploadTask.snapshot.ref;
              const fileName = ref.name;

              ref
                .getDownloadURL()
                .then(downloadURL => {
                  startUpdateDownloadURL(downloadURL, fileName);
                })
                .then(() => {
                  setError('');
                  setUploading(false);
                  e.target.reset();
                });
            }
          );
        });
    } else setError('No file selected');
  };

  return (
    <form onSubmit={fileUpload}>
      {error && <p>{error}</p>}
      <input
        type="file"
        id={id}
        onClick={e => e.stopPropagation()}
        accept=".pdf"
        disabled={uploading}
      />
      <button onClick={e => e.stopPropagation()} disabled={uploading}>
        {uploading ? 'Uploading...' : 'Upload'}
      </button>
    </form>
  );
};

const mapStateToProps = ({ tasks, auth }, { id }) => {
  const uid = auth.user.uid;
  const task = tasks.find(task => task.id === id);

  return {
    TaskEvent: firebase.storage.TaskEvent,
    fileExists: task ? task.downloadURLs.hasOwnProperty(uid) : false,
    uid
  };
};

const mapDispatchToProps = (dispatch, { id, gid, namingConvention }) => ({
  uploadFile: file => dispatch(uploadFile(file, id, namingConvention)),
  startUpdateDownloadURL: (downloadURL, fileName) =>
    dispatch(startUpdateDownloadURL(id, gid, downloadURL, fileName)),
  startRemoveUserFile: uid => dispatch(startRemoveUserFile(id, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(FileUploadForm);
