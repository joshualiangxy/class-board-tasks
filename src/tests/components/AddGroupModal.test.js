import React from 'react';
import { shallow } from 'enzyme';
import { AddGroupModal } from '../../components/AddGroupModal';
import { groupSetOne } from '../fixtures/groups';

const onRequestClose = jest.fn();
const preventDefault = jest.fn();
const startNewGroup = jest.fn((submittedName, submittedMod) =>
  Promise.resolve()
);

const testOne = {
  groupName: 'testTwo',
  module: 'testMod'
};

const testTwo = {
  groupName: 'testTwo',
  module: ''
};

const testThree = {
  groupName: ' ',
  module: 'testThree'
};

const testFour = {
  groupName: '',
  module: ''
};
let wrapper;

beforeEach(() => {
  jest.clearAllMocks();
  wrapper = shallow(
    <AddGroupModal
      isOpen={true}
      onRequestClose={onRequestClose}
      startNewGroup={startNewGroup}
    />
  );
});

describe('render', () => {
  it('should render AddGroupModal', () => expect(wrapper).toMatchSnapshot());

  it('should have a group name and module input', () => {
    expect(wrapper.find('input[type="text"]')).toHaveLength(2);
  });

  it('should render with placeholder text', () => {
    expect(wrapper.find('input').at(0).prop('placeholder')).toEqual(
      'Group name (required)'
    );
    expect(wrapper.find('input').at(1).prop('placeholder')).toEqual(
      'Module code/Class name (Optional)'
    );
  });
});

describe('input', () => {
  it('should have input change for groupName', () => {
    expect(wrapper.find('input').at(0).prop('value')).toEqual('');
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testOne.groupName }
    });
    expect(wrapper.find('input').at(0).prop('value')).toEqual(
      testOne.groupName
    );
  });

  it('should have input change for module', () => {
    expect(wrapper.find('input').at(1).prop('value')).toEqual('');
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testOne.module } }
    });
    expect(wrapper.find('input').at(1).prop('value')).toEqual(testOne.module);
  });
});

describe('submit', () => {
  it('should start groups with name and module', () => {
    // Set input
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testOne.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testOne.module } }
    });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(startNewGroup).toHaveBeenCalledTimes(1);
    expect(startNewGroup).toHaveBeenLastCalledWith(
      testOne.groupName,
      testOne.module
    );
  });

  it('should start groups with name and no module', () => {
    // Set input
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testTwo.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testTwo.module } }
    });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(startNewGroup).toHaveBeenCalledTimes(1);
    expect(startNewGroup).toHaveBeenLastCalledWith(
      testTwo.groupName,
      testTwo.module
    );
  });

  it('should not start groups with whitespace name with valid module', () => {
    // Set input
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testThree.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testThree.module } }
    });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(startNewGroup).toHaveBeenCalledTimes(0);
    expect(wrapper.find('.form__error').text()).toEqual(
      'Please enter a group name'
    );
  });

  it('should not start groups with no name and no module', () => {
    // Set input
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testFour.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testFour.module } }
    });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(startNewGroup).toHaveBeenCalledTimes(0);
    expect(wrapper.find('.form__error').text()).toEqual(
      'Please enter a group name'
    );
  });

  it('should reset error message on success', () => {
    // Set input
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testFour.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testFour.module } }
    });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault });
    // Set input to working group
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testTwo.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testTwo.module } }
    });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault });
    expect(wrapper.find('.form__error')).toHaveLength(0);
  });
});

describe('cancel', () => {
  it('should reset all fields on cancel', () => {
    // Set input
    wrapper.find('input').at(0).prop('onChange')({
      target: { value: testOne.groupName }
    });
    wrapper.find('input').at(1).prop('onChange')({
      target: { value: { toUpperCase: () => testOne.module } }
    });
    wrapper.find('Modal').at(0).prop('onRequestClose')();

    expect(onRequestClose).toHaveBeenCalledTimes(1);
    expect(wrapper.find('input').at(0).prop('value')).toEqual('');
    expect(wrapper.find('input').at(1).prop('value')).toEqual('');
    expect(wrapper.find('.form__error')).toHaveLength(0)
  });
});
