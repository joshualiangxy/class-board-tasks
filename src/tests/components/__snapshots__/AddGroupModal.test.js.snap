// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render AddGroupModal 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={200}
  contentLabel="New Group"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Group name (required)"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Module code/Class name (Optional)"
      type="text"
      value=""
    />
    <button>
      Submit
    </button>
    <button
      onClick={[Function]}
    >
      Cancel
    </button>
    <div
      className="error"
    />
  </form>
</Modal>
`;
